File: scala_maze/src/main/scala/Maze.scala
package Maze

import scala.collection.mutable.ListBuffer

enum Exploration:
  case Explored, UnExplored

enum Maze:
  case Branch(
      label: String,
      left: Maze,
      right: Maze,
      var status: Exploration = Exploration.UnExplored
  )
  case Leaf(label: String, var status: Exploration = Exploration.UnExplored)
  def explore(): List[String] =
    val buf = scala.collection.mutable.ListBuffer.empty[String]
    this.explore(buf)
    buf.toList
  def explore(trace: ListBuffer[String]): Unit =
    println(trace)
    this match
      case branch @ Branch(label, left, right, status) =>
        status match
          case Exploration.UnExplored =>
            branch.status = Exploration.Explored;
            trace += label;
            left.explore(trace);
            right.explore(trace);
          case Exploration.Explored => return
      case leaf @ Leaf(label, status) =>
        status match
          case Exploration.UnExplored =>
            leaf.status = Exploration.Explored;
            trace += label;
          case Exploration.Explored => return

File: rust_maze/src/maze.rs
use std::fmt;
use std::sync::Arc;
use std::sync::Mutex;

#[derive(PartialEq, Eq, Debug)]
pub enum Exploration {
    Explored,
    LeftExplored,
    Unexplored,
}

pub enum Maze {
    Branch {
        label: String,
        left: Arc<Maze>,
        right: Arc<Maze>,
        status: Mutex<Exploration>,
    },
    Leaf {
        label: String,
        status: Mutex<Exploration>,
    },
}

impl fmt::Debug for Maze {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Maze::Branch { label, .. } => {
                write!(f, "Branch: {}", label)
            }
            Maze::Leaf { label, .. } => {
                write!(f, "Leaf: {}", label)
            }
        }
    }
}

impl Maze {
    pub fn explore(
        &self,
        copy: Arc<Maze>,
        stack: Arc<Mutex<Vec<Arc<Maze>>>>,
        trace: Arc<Mutex<Vec<Arc<Maze>>>>,
    ) {
        match self {
            Maze::Branch {
                left,
                right,
                status,
                ..
            } => {
                let mut status_uw = status.lock().unwrap();
                let mut stack_uw = stack.lock().unwrap();
                let mut trace_uw = trace.lock().unwrap();
                if *status_uw == Exploration::Unexplored {
                    *status_uw = Exploration::LeftExplored;
                    trace_uw.push(Arc::clone(&copy));
                    stack_uw.push(Arc::clone(&copy));
                    stack_uw.push(Arc::clone(left));
                } else if *status_uw == Exploration::LeftExplored {
                    *status_uw = Exploration::Explored;
                    stack_uw.push(Arc::clone(right));
                }
            }
            Maze::Leaf { status, .. } => {
                let mut status_uw = status.lock().unwrap();
                let mut trace_uw = trace.lock().unwrap();
                if *status_uw == Exploration::Unexplored {
                    *status_uw = Exploration::Explored;
                    trace_uw.push(copy);
                }
            }
        }
    }
}
